pkg/deploy/identity_provider.go

```diff
@@ -0,0 +1,265 @@
+package deploy
+
+import (
+	"context"
+	"strings"
+
+	"github.com/eclipse/che-operator/pkg/util"
+	oauth "github.com/openshift/api/oauth/v1"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// SyncIdentityProviderToCluster instantiates the identity provider (Keycloak) in the cluster. Returns true if
+// the provisioning is complete, false if requeue of the reconcile request is needed.
+func SyncIdentityProviderToCluster(deployContext *DeployContext, cheHost string, protocol string, cheFlavor string) (bool, error) {
+	instance := deployContext.CheCluster
+	ingressStrategy := util.GetServerExposureStrategy(instance, DefaultServerExposureStrategy)
+	ingressDomain := instance.Spec.K8s.IngressDomain
+	cheMultiUser := GetCheMultiUser(instance)
+	tests := util.IsTestMode()
+	isOpenShift := util.IsOpenShift
+
+	if instance.Spec.Auth.ExternalIdentityProvider {
+		return true, nil
+	}
+
+	if cheMultiUser == "false" {
+		if util.K8sclient.IsDeploymentExists("keycloak", instance.Namespace) {
+			util.K8sclient.DeleteDeployment("keycloak", instance.Namespace)
+		}
+
+		return true, nil
+	}
+
+	keycloakLabels := GetLabels(instance, "keycloak")
+
+	serviceStatus := SyncServiceToCluster(deployContext, "keycloak", []string{"http"}, []int32{8080}, keycloakLabels)
+	if !tests {
+		if !serviceStatus.Continue {
+			logrus.Info("Waiting on service 'keycloak' to be ready")
+			if serviceStatus.Err != nil {
+				logrus.Error(serviceStatus.Err)
+			}
+
+			return false, serviceStatus.Err
+		}
+	}
+
+	exposureStrategy := util.GetServerExposureStrategy(instance, DefaultServerExposureStrategy)
+	singleHostExposureType := GetSingleHostExposureType(instance)
+	useGateway := exposureStrategy == "single-host" && (util.IsOpenShift || singleHostExposureType == "gateway")
+
+	// create Keycloak ingresses when on k8s
+	var keycloakURL string
+	if !isOpenShift {
+		if useGateway {
+			// try to guess where in the ingress-creating code the /auth endpoint is defined...
+			cfg := GetGatewayRouteConfig(instance, "keycloak", "/auth", 10, "http://keycloak:8080", false)
```

Comment:
you've named all gateway routes configmaps with `-gtw` suffix, but keycloak one is named only `keycloak`. Sync it please so it's more clear.

Reply:
Good point. Renamed them all to `che-gateway-route-*`.